server/application/application.go

```diff
@@ -82,7 +83,17 @@ func (s *Server) Create(ctx context.Context, a *appv1.Application) (*appv1.Appli
 	if err != nil {
 		return nil, err
 	}
-	return s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Create(a)
+	out, err := s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Create(a)
+	if apierr.IsAlreadyExists(err) {
+		// act idempotent if existing spec matches new spec
+		existing, err2 := s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Get(a.Name, metav1.GetOptions{})
+		if err2 == nil {
+			if reflect.DeepEqual(existing.Spec, a.Spec) {
```

Comment:
DeepEqual works right now, but in the near future, the there will be parts of the spec which gets set independently from the creation process. For example, when a user requests to perform an application operation, it will be specified in the .spec.operation field. This would fail the DeepEqual check. I think to be safe, this should be changed to only DeepEqual .Spec.Source and .Spec.Destination

Reply:
Talked to @alexmt, we will put operation as a peer to spec and status. So your implementation is fine.